### PR TITLE
Add cluster operations components

### PR DESCRIPTION
**Pull request by corindwyer**
### Description of the Change
Refactor the existing agent removal process into a new component `cluster operations`.

### CODE REVIEW

File: titus-server-master/src/main/java/com/netflix/titus/master/clusteroperations/ClusterAgentRemover.java [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+            if (totalEligibleAgentsToRemove <= 0) {
+                return Observable.empty();
+            }
+
+            long maxTokensToTake = Math.min(TOKEN_BUCKET_CAPACITY, totalEligibleAgentsToRemove);
+            Optional<Pair<Long, ImmutableTokenBucket>> takeOpt = ClusterAgentRemover.this.lastTokenBucket.tryTake(1, maxTokensToTake);
+            if (takeOpt.isPresent()) {
+                Pair<Long, ImmutableTokenBucket> takePair = takeOpt.get();
+                ClusterAgentRemover.this.lastTokenBucket = takePair.getRight();
+                int tokensAvailable = Math.toIntExact(takePair.getLeft());
```
[the remark this answers is not in the document]
Comment (corindwyer) [reply titus-server-master/src/main/java/com/netflix/titus/master/clusteroperations/ClusterAgentRemover.java]: will change to use longs and get rid of this

File: titus-server-master/src/main/java/com/netflix/titus/master/clusteroperations/ClusterAgentRemover.java [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+                for (Map.Entry<AgentInstanceGroup, List<AgentInstance>> entry : eligibleAgentInstancesPerInstanceGroup.entrySet()) {
+                    int tokensRemaining = tokensAvailable - tokensUsed;
+                    if (tokensRemaining <= 0) {
+                        break;
+                    }
+                    String instanceGroupId = entry.getKey().getId();
+                    List<AgentInstance> agentInstances = entry.getValue();
+                    List<AgentInstance> agentInstancesToTerminate = agentInstances.size() > tokensAvailable ? agentInstances.subList(0, tokensRemaining) : agentInstances;
+                    List<String> agentInstanceIdsToTerminate = agentInstancesToTerminate.stream().map(AgentInstance::getId).collect(Collectors.toList());
+                    if (agentInstanceIdsToTerminate.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (corindwyer) [reply titus-server-master/src/main/java/com/netflix/titus/master/clusteroperations/ClusterAgentRemover.java]: fixed

File: titus-server-master/src/main/java/com/netflix/titus/master/clusteroperations/ClusterAgentRemover.java [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+                        continue;
+                    }
+                    logger.info("Terminating agent instances: {} for instance group: {}", agentInstanceIdsToTerminate, instanceGroupId);
```
[the remark this answers is not in the document]
Comment (corindwyer) [reply titus-server-master/src/main/java/com/netflix/titus/master/clusteroperations/ClusterAgentRemover.java]: fixed

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/constraint/AgentManagementConstraint.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        InstanceOverrideState overrideState = instance.getOverrideStatus().getState();
+        if (overrideState != InstanceOverrideState.None) {
+            return INSTANCE_STATE_IS_OVERRIDDEN;
+        }
+
+        if (!agentStatusMonitor.isHealthy(instanceId)) {
+            return INSTANCE_UNHEALTHY;
+        }
+
+        return null;
```
[the remark this answers is not in the document]
Comment (corindwyer) [reply titus-server-master/src/main/java/com/netflix/titus/master/scheduler/constraint/AgentManagementConstraint.java]: fixed

File: titus-server-master/src/main/java/com/netflix/titus/master/clusteroperations/ClusterAgentRemover.java [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+                        if (resultItem.hasValue()) {
+                            if (resultItem.getValue()) {
+                                terminatedOk.add(instanceId);
+                            } else {
+                                errors.put(instanceId, "Terminate status 'false'");
+                            }
+                        } else {
+                            errors.put(instanceId, resultItem.getError().getMessage());
+                        }
+                        agentInstancesBeingTerminated.invalidate(instanceId);
```
[the remark this answers is not in the document]
Comment (corindwyer) [reply titus-server-master/src/main/java/com/netflix/titus/master/clusteroperations/ClusterAgentRemover.java]: Why is this too early? At this point, there is an known error for this instanceId action and it is not going to change right?

File: titus-server-master/src/main/java/com/netflix/titus/master/clusteroperations/ClusterAgentRemover.java [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+                TOKEN_BUCKET_REFILL_INTERVAL_MS, TimeUnit.MILLISECONDS);
+        this.lastTokenBucket = ImmutableLimiters.tokenBucket(TOKEN_BUCKET_CAPACITY, immutableRefillStrategy);
+    }
+
+    private Observable<List<Either<Boolean, Throwable>>> createTerminateAction(String instanceGroupId, List<String> terminateIds) {
+        Stopwatch timer = Stopwatch.createStarted();
+        terminateIds.forEach(terminateId -> agentInstancesBeingTerminated.put(terminateId, terminateId));
+        return agentManagementService.terminateAgents(instanceGroupId, terminateIds, true)
+                .doOnNext(result -> {
+                    if (result.size() != terminateIds.size()) {
```
[the remark this answers is not in the document]
Comment (corindwyer) [reply titus-server-master/src/main/java/com/netflix/titus/master/clusteroperations/ClusterAgentRemover.java]: fixed

File: titus-server-master/src/main/java/com/netflix/titus/master/clusteroperations/ClusterAgentRemover.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (tbak) [reply titus-server-master/src/main/java/com/netflix/titus/master/clusteroperations/ClusterAgentRemover.java]:
```  if (resultItem.hasValue()) {
+                            if (resultItem.getValue()) {
+                                terminatedOk.add(instanceId);
+                            } else {
+                                errors.put(instanceId, "Terminate status 'false'");
+                            }
+                        } else {
+                            errors.put(instanceId, resultItem.getError().getMessage());
+                        }
+                        agentInstancesBeingTerminated.invalidate(instanceId);```

The result (success or failure) is from AWS API, so the instance is in the process of being terminated. I thought that `agentInstancesBeingTerminated` keeps these instances for some time (5-10min), until the termination action happens. If I understand this code correctly, now the id is add before AWS call, and immediately removed after it completes.

[the remark this answers is not in the document]
Comment (corindwyer) [reply titus-server-master/src/main/java/com/netflix/titus/master/clusteroperations/ClusterAgentRemover.java]: I agree. Going to remove this invalidation and all requests will need to wait on cache ttl time for removal.

**Pull request merged**